fiona/_geometry.pyx

```diff
@@ -143,20 +146,49 @@ cdef class GeomBuilder:
 
     cdef build(self, void *geom):
         # The only method anyone needs to call
-        if geom == NULL:
-            raise ValueError("Null geom")
-
-        cdef unsigned int etype = OGR_G_GetGeometryType(geom)
+        cdef void *cogr_geometry = NULL
+        cdef void *cloned_geometry = NULL
 
-        self.code = base_geometry_type_code(etype)
+        if geom == NULL:
+            return None
+
+        code = base_geometry_type_code(OGR_G_GetGeometryType(geom))
+
+        # We convert special geometries (curves, TIN, Triangle, ...)
+        # to GeoJSON compatible geometries (LineStrings, Polygons, MultiPolygon, ...)
+        if 8 <= code <= 14:  # Curves.
+            # OGR_G_GetLinearGeometry: The ownership of the returned geometry belongs to the caller.
+            cogr_geometry = OGR_G_GetLinearGeometry(geom, 0.0, NULL)
+            code = base_geometry_type_code(OGR_G_GetGeometryType(cogr_geometry))
+            ndims = OGR_G_GetCoordinateDimension(cogr_geometry)
+            geom = cogr_geometry
+        elif 15 <= code <= 17:
+            cloned_geometry = OGR_G_Clone(geom)
+            if code in (15, 16):
+                # OGR_G_ForceToMultiPolygon: the converted geometry (ownership to caller).
+                cogr_geometry = OGR_G_ForceToMultiPolygon(cloned_geometry)
+            elif code == 17:
+                # OGR_G_ForceToPolygon: the converted geometry (ownership to caller).
+                cogr_geometry = OGR_G_ForceToPolygon(cloned_geometry)
+            code = base_geometry_type_code(OGR_G_GetGeometryType(cogr_geometry))
+            ndims = OGR_G_GetCoordinateDimension(cogr_geometry)
+            geom = cogr_geometry
+        else:
+            ndims = OGR_G_GetCoordinateDimension(geom)
```

Comment:
ndims = OGR_G_GetCoordinateDimension(geom) could be moved outside all branches of the if / elif / else blocks.

Reply:
True. Neither OGR_G_GetLinearGeometry, OGR_G_ForceToMultiPolygon or OGR_G_ForceToPolygon should have an influence on the number of dimensions.

Reply:
I meant that this call is repeated in the 3 branches, and could be just done once after them